src/facade/ether-anchor-earn.ts

```diff
@@ -0,0 +1,637 @@
+                  atokenBalance.mul(exchangeRate).div(constants.WeiPerEther),
+                ),
+                APY: utils.formatEther(apy),
+              };
+            }
+          },
+        ),
+      );
+
+      const blockNumber = await this._provider.getBlockNumber();
+
+      return new MarketOutput(
+        this.chain,
+        this.network.toString(),
+        blockNumber,
+        markets,
+      );
+    }
+
+    private async waitForTx(txHash: string): Promise<void> {
+      await this._provider.waitForTransaction(txHash, this._confirmation);
+    }
+
+    private accumulateTxFee(txs: PopulatedTransaction[]): string {
+      return txs
+        .reduce(
+          (acc, tx) => acc.add(tx.gasLimit.mul(tx.gasPrice)),
+          BigNumber.from(0),
+        )
+        .toString();
+    }
+
+    private toOutput<
+      T extends CustomSigner<UnsignedTx, SignedTx> &
+        CustomBroadcaster<UnsignedTx, SignedTx> &
+        Loggable<Output | OperationError> & { currency: Denoms; amount: string }
+    >(
+      { status, type, txDetails = [], txFee }: OutputFragment,
+      { currency, amount }: T,
+    ): Output {
+      return {
+        chain: this.chain,
+        network: this.network,
+        status,
+        type,
+        currency,
+        amount,
+        txDetails,
+        txFee,
+      };
+    }
+
+    private toError(txType: TxType, e: Error): OperationError {
+      return {
+        chain: this.chain,
+        network: this.network,
+        status: STATUS.UNSUCCESSFUL,
+        type: txType,
+        error_msg: e.message,
+      };
+    }
+
+    private respToDetail(txResp: providers.TransactionResponse): TxDetails {
+      return {
+        chain: this.chain,
+        network: this.network,
+        height: txResp.blockNumber,
+        timestamp: new Date(txResp.timestamp * 1000),
+        txHash: txResp.hash,
+      };
+    }
+
+    private async signAndBroadcast<
+      T extends CustomSigner<UnsignedTx, SignedTx> &
+        CustomBroadcaster<UnsignedTx, SignedTx> &
+        Loggable<Output | OperationError> & { currency: Denoms; amount: string }
+    >(tx: PopulatedTransaction, txType: TxType, option: T): Promise<TxDetails> {
+      const { customSigner, customBroadcaster } = option;
+      const logger = option.log || (() => void 0);
+
+      if (!(customSigner && customBroadcaster) && this._wallet === null) {
+        throw new Error('environment not satisfied');
+      }
+
+      try {
+        if (customBroadcaster) {
+          const txHash = await customBroadcaster(tx);
+          const txResp = await this._provider.getTransaction(txHash);
+          return this.respToDetail(txResp);
+        }
+
+        const signer =
+          customSigner ||
+          ((tx: providers.TransactionRequest) =>
+            this._wallet.signTransaction(tx));
+
+        const signedTx = await signer(tx);
+        const txResp = await this._provider.sendTransaction(signedTx);
+        logger(
```

Comment:
Please add a stateful logger so that a user can track the current state of the tx.